modules/libcom/src/osi/osiFileName.c

```diff
@@ -0,0 +1,197 @@
+// cf. https://docs.microsoft.com/en-us/dotnet/standard/io/file-path-formats#canonicalize-separators
+#  define SEP(C) (C=='/' || C=='\\')
+#else
+#  define SEP(C) (C=='/')
+#endif
+
+static const char magicCWD[] = "<current_directory>";
+const char* const epicsPathJoinCurDir = magicCWD;
+
+static
+size_t epicsPathDF(char *path, size_t pathlen, int outfile)
+{
+    size_t pos;
+
+    /* search forward to ignore unused buffer */
+    pathlen = epicsStrnLen(path, pathlen);
+
+    /* search backwards for last separator */
+    pos = pathlen;
+    for(; pos>0; pos--) {
+        char c = path[pos-1u];
+        if(SEP(c)) {
+            /* found separator */
+            if(outfile) {
+                memcpy(path, &path[pos], pathlen-pos);
+                pos = pathlen-pos;
+            }
+            goto done;
+        }
+    }
+    /* no separator found */
+    if(outfile) {
+        pos = pathlen;
+    }
+
+done:
+    if(pos < pathlen) {
+        path[pos] = '\0';
+    }
+
+    return pos;
+}
+
+size_t epicsPathDir(char *path, size_t pathlen)
+{
+    return epicsPathDF(path, pathlen, 0);
+}
+
+size_t epicsPathFile(char *path, size_t pathlen)
+{
+    return epicsPathDF(path, pathlen, 1);
+}
+
+int epicsPathIsAbs(const char *path, size_t pathlen)
+{
+    /* for *NIX this is simple.  path[0]=='/' is absolute.  Anything else is relative
+     *
+     * for windows... not so much.
+     * cf. https://docs.microsoft.com/en-us/dotnet/standard/io/file-path-formats#identify-the-path
+     *
+     * we treat as absolute:
+     *  - device and UNC paths.
+     *  - absolute on current drive.  eg. "\foo" -> "C:\foo" if %CD% starts with C:
+     *  - absolute w/ drive.  eg. "C:\foo"
+     *
+     * we treat as relative
+     *  - relative to some drive.  eg. "C:foo" -> "D:\bar\foo" if %CD% == "D:\bar"  (sneaky!)
+     *  - anything else
+     */
+    if(pathlen>0 && SEP(path[0])) // *NIX: simple abs.  windows: device path, UNC path, or root of current drive
+        return 1;
+#if defined(_WIN32) || defined(__CYGWIN__)
+    if(pathlen>=3 && ((path[0]>='a' && path[0]<='z') || (path[0]>='A' && path[0]<='Z')) && path[1]==':' && SEP(path[2])) {
+        /* eg. C:FOO is relative
+         *     C:/FOO is absolute
+         */
+        return 1;
+    }
+#endif
+    return 0;
+}
+
+char* epicsPathJoin(const char * const *fragments, size_t nfragments)
+{
+    char* ret;
+    char* cwd = NULL;
+    size_t i;
+    size_t firstfrag = 0;
+    size_t nfrags = 0; /* nfragments excluding NULL and "" */
+    size_t totallen = 0;
+
+    STATIC_ASSERT(sizeof(OSI_PATH_SEPARATOR)==2);
+
+    /* first pass to compute buffer size */
+    for(i=0; i<nfragments; i++) {
+        const char* frag = fragments[i];
+        size_t flen;
+        if(!frag || frag[0]=='\0')
+            continue;
+
+        if(frag==magicCWD) {
+            if(!cwd)
+                cwd = epicsPathAllocCWD();
+            if(!cwd)
+                return NULL;
+            frag = cwd;
+        }
+
```

Comment:
If the process current directory was `c:\abc` and it's previous directory on d: was `d:\efg` then if it was passed ("d:", "<current_directory>") should it generate `c:\abc` or `d:\efg` ? I don't know how likely this is and whether we need to cover it, but we would need to use  `getdcwd` rather than `getcwd`

Reply:
Assuming that the current directory (`getcwd()`) is always an absolute path, a join will ignore any prior fragments.  So `epicsPathJoin(["d:", epicsPathJoinCurDir, "foo"], 3)` will ignore `"d:"`.

Reply:
Also, `epicsPathJoinCurDir` holds a magic address which must be used.  The fact that this address prints as `"<current_directory>"` is a convenience.

Reply:
Ah ok, we probably don't need to worry about windows having a drive specific current working directory then